beacon/light/request/scheduler.go

```diff
@@ -0,0 +1,371 @@
+	for _, module := range s.modules {
+		log.Debug("Processing module", "name", s.names[module], "request events", len(s.requestEvents[module]))
+		module.Process(append(serverEvents, requestEvents[module]...))
+	}
+}
+
+func (s *Scheduler) sendRequests() {
+	servers := make(serverSet)
+	for server := range s.servers {
+		if ok, _ := server.canRequestNow(); ok {
+			servers[server] = struct{}{}
+		}
+	}
+	log.Debug("Generating request candidates", "servers", len(servers))
+
+	for _, module := range s.modules {
+		if len(servers) == 0 {
+			return
+		}
+		s.eventLock.Lock() // ensure that EvRequest is added first
+		if req, sent := s.tryRequest(module, servers); sent {
+			log.Debug("Sent request", "module", s.names[module])
+			s.addEvent(Event{
+				Type:   EvRequest,
+				Server: req.Server,
+				Data: RequestResponse{
+					ID:      req.ID,
+					Request: req.Request,
+				},
+			})
+		}
+		s.eventLock.Unlock()
+	}
+}
+
+func (s *Scheduler) tryRequest(module Module, servers serverSet) (RequestWithID, bool) {
+	var (
+		maxServerPriority, maxRequestPriority float32
+		bestServer                            server
+		bestRequest                           Request
+	)
+	maxServerPriority, maxRequestPriority = -math.MaxFloat32, -math.MaxFloat32
+	serverCount := len(servers)
+	var removed, candidates int
+	for server := range servers {
+		canRequest, serverPriority := server.canRequestNow()
+		if !canRequest {
+			delete(servers, server)
+			removed++
+			continue
+		}
+		request, requestPriority := module.MakeRequest(server)
+		if request != nil {
+			candidates++
+		}
+		if request == nil || requestPriority < maxRequestPriority ||
+			(requestPriority == maxRequestPriority && serverPriority <= maxServerPriority) {
+			continue
+		}
+		maxServerPriority, maxRequestPriority = serverPriority, requestPriority
+		bestServer, bestRequest = server, request
+	}
+	log.Debug("Request attempt", "serverCount", serverCount, "removedServers", removed, "requestCandidates", candidates)
+	if bestServer == nil {
+		return RequestWithID{}, false
+	}
+	id := ServerAndID{Server: bestServer, ID: bestServer.sendRequest(bestRequest)}
+	s.pending[id] = pendingRequest{request: bestRequest, module: module}
+	return RequestWithID{ServerAndID: id, Request: bestRequest}, true
+}
+
+// Trigger starts a new processing round. If fired during processing, it ensures
+// another full round of processing all modules.
+func (s *Scheduler) Trigger() {
+	select {
+	case s.triggerCh <- struct{}{}:
+	default:
+	}
+}
+
+// addEvent adds an Event either as a request event or a server event, depending
+// on its type. In case of an EvUnregistered server event it also closes all
+// pending requests to the given server by adding a failed request event (EvFail),
+// ensuring that all requests get finalized and thereby allowing the module logic
+// to be safe and simple.
+func (s *Scheduler) addEvent(event Event) {
+	if _, ok := s.servers[event.Server.(server)]; !ok {
```

Comment:
Done. Though it really should not happen, the interface is only used because of the module tests which do not use `Scheduler`, but now it prints an error log instead of a panic if it happens for some reason.